## 1. Random Forest with Fingerprint

In this baseline model, we want to use compound's fingerprint to predict the outcome of assay. This is a single task, and the prediction performance is measured by the test set with cross-validation.

The fingerprint is extracted from all compounds with length `1024`.

In [6]:
import numpy as np
import pandas as pd
import pickle
from json import load, dump
from sklearn.ensemble import RandomForestClassifier
from collections import Counter
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import cross_validate, StratifiedKFold, GridSearchCV
from sklearn.linear_model import LogisticRegression

In [9]:
# Load the output matrix
output_matrix_npz = np.load("./resource/output_matrix.npz")
compound_inchi = output_matrix_npz["compound_inchi"]
compound_broad_id = output_matrix_npz["compound_broad_id"]
output_matrix = output_matrix_npz["output_matrix"]
output_matrix.shape

(9404, 141)

In [4]:
# Load the fingerprint
fingerprint = np.load("./resource/fp_features.npz")

# Build the feature array
fp_feature = np.zeros((len(compound_broad_id), fingerprint["features"].shape[1]))

# Choose selected compounds and arrange them by the output matrix order
fp_index = dict(zip(fingerprint["names"].astype(str), range(len(fingerprint["names"]))))

for b in range(len(compound_broad_id)):
    bid = compound_broad_id[b]
    fp_feature[b, :] = fingerprint["features"][fp_index[bid], :]

It took one hour to create this $9404 \times 1024$ feature matrix with finger print on condor.

In [8]:
fp_feature = np.load("./resource/extracted_fp_feature.npz")["feature"]
fp_feature.shape

(9404, 1024)

From the PriA-SSB study, some good parameters for a random forests are selected with a different assay dataset. We expect the best parameters (`RF_h`) to perform well in this dataset.

For each assay, we only train and predict on compounds that give non-NA results.

![](https://user-images.githubusercontent.com/3532898/46812713-e8a93280-cd3a-11e8-949a-3469236a3943.png)

In [74]:
def train_rf_on_assay(assay_array, rf, kfold=5, n_jobs=4):
    """
    Train and measure a random forest model in a cross validation fasshion.
    The model is only trained on compounds which give non-NA results.
    """
    # Filter out NA from the assay_array
    y_index = [False if a == -1 else True for a in assay_array ]
    y = assay_array[y_index]
    x = fp_feature[y_index, :]
    
    if Counter(y)[1] < kfold:
        print("Warning: the number of total postives is less than kfold")
        
    # One-hot encode y array
    #y = np.vstack([[1, 0] if i == 1 else [0, 1] for i in y])
    
    # Build the cross validation scheme
    # Since some assays have extremely small number of positives,
    # I will use StratifiedKFold to preserve the proportion of positive in test set
    
    my_kfold_gen = StratifiedKFold(n_splits=kfold, shuffle=True)
    scoring = ["f1", "accuracy", "precision", "recall", "average_precision", "roc_auc"]
    cv = cross_validate(rf, x, y, scoring=scoring, cv=my_kfold_gen, n_jobs=n_jobs,
                        return_train_score=False)
    
    cv["total_count"] = Counter(y)
    
    return cv

In [86]:
results = {}

rf_classifier = RandomForestClassifier(n_estimators=8000, max_features="log2",
                                       min_samples_leaf=1, class_weight="balanced")

for i in range(output_matrix.shape[1]):
    print(i)
    results[i] = train_rf_on_assay(output_matrix[:,i], rf_classifier, kfold=4)

We do a stratified 5-fold cross validation for each assay and record metrics on 5 test sets.

It took 2 hour to train these 141 single tasks. We can visualize the results.

In [8]:
results = np.load("./resource/results.npz")['results'].item()

In [6]:
def get_mean_score(results):
    """
    Aggregate each score over 5 test sets.
    """
    mean_df = {
        "f1": [],
        "accuracy": [],
        "average_precision": [],
        "roc_auc": [],
        "precision": [],
        "recall": [],
        "pos_num": [],
        "neg_num": []
    }
    
    for k, r in results.items():
        mean_df["f1"].append(np.mean(r["test_f1"]))
        mean_df["accuracy"].append(np.mean(r["test_accuracy"]))
        mean_df["average_precision"].append(np.mean(r["test_average_precision"]))
        mean_df["roc_auc"].append(np.mean(r["test_roc_auc"]))
        mean_df["precision"].append(np.mean(r["test_precision"]))
        mean_df["recall"].append(np.mean(r["test_recall"]))
        mean_df["pos_num"].append(r["total_count"][1])
        mean_df["neg_num"].append(r["total_count"][0])
    
    return pd.DataFrame(mean_df)


In [9]:
mean_df = get_mean_score(results)
mean_df.to_csv("./mean_df.csv", index=False)
mean_df.head(10)

f1  accuracy  average_precision   roc_auc  precision    recall  \
0  0.080000  0.989659           0.249031  0.753840   0.200000  0.050000   
1  0.759369  0.682051           0.781926  0.746656   0.676644  0.867821   
2  0.183729  0.781891           0.520559  0.776232   0.533333  0.121905   
3  0.158615  0.828373           0.400695  0.698030   0.805556  0.090476   
4  0.030199  0.930673           0.168273  0.654066   0.400000  0.015692   
5  0.660623  0.701898           0.792594  0.774347   0.704681  0.622222   
6  0.000000  0.992296           0.026350  0.657967   0.000000  0.000000   
7  0.580000  0.470000           0.777778  0.633333   0.500000  0.733333   
8  0.250102  0.678020           0.509820  0.675496   0.553435  0.162791   
9  0.019755  0.886010           0.264240  0.685354   0.566667  0.010095   

   pos_num  neg_num  
0       17     1530  
1      196      143  
2       74      256  
3      210      932  
4      128     1747  
5      180      206  
6       24     3221  
7       11       10  
8      215      431  
9      396     3078

In [37]:
print(("Across 141 assays, the average f1={:.2f}%, accuracy={:.2f}%, ap={:.2f}%, auc={:.2f}%, " + \
       "precision={:.2f}%, recall={:.2f}%.").format(
    np.mean(mean_df["f1"]) * 100,
    np.mean(mean_df["accuracy"]) * 100,
    np.mean(mean_df["average_precision"]) * 100,
    np.mean(mean_df["roc_auc"]) * 100,
    np.mean(mean_df["precision"]) * 100,
    np.mean(mean_df["recall"]) * 100))

Across 141 assays, the average f1=10.03%, accuracy=91.96%, ap=30.35%, auc=72.19%, precision=27.76%, recall=7.83%.


![random_forest_plot_1.png](./plots/random_forest_plot_1.png)

### 1.1. Comments

- Accuracies are pretty high except some outliers with few assays. It is due to the skewness of positive samples.
- This baseline model performs poorly based on metrics `F1`, `AP`, `Precision`, `Recall`.
- However, this model has high `AUC`, which is the main metric used in the ICCR paper.
- Using `AUC`, finterprint baseline gives a similar result as ICCR paper's advanced model (attached below).
- `AP` has a very large variance. Low sample size (both pos and neg) tends to give high `AP` values.

![](https://i.imgur.com/Z790iMB.png)

### 1.2. Random Forest + Finger Print using Converted InChI

I have built an output matrix for the intersected compounds using converted InChI strings. I have also excluded some problematic entries (conversion collision) from the intersection.

This output matrix uses 209 assays (increasing from 141), and 26638 compounds (increasing from 9404). It should give us a more fair comparison.

In [39]:
# Load the output matrix
output_matrix_npz = np.load("./resource/output_matrix_convert.npz")
compound_inchi = output_matrix_npz["compound_inchi"]
compound_broad_id = output_matrix_npz["compound_broad_id"]
output_matrix = output_matrix_npz["output_matrix"]
output_matrix.shape

(26638, 209)

In [40]:
# Load the fingerprint
fingerprint = np.load("./resource/fp_features.npz")

# Build the feature array
fp_feature = np.zeros((len(compound_broad_id), fingerprint["features"].shape[1]))

# Choose selected compounds and arrange them by the output matrix order
fp_index = dict(zip(fingerprint["names"].astype(str), range(len(fingerprint["names"]))))

for b in range(len(compound_broad_id)):
    bid = compound_broad_id[b]
    fp_feature[b, :] = fingerprint["features"][fp_index[bid], :]

In [12]:
results = np.load("./resource/results_converted.npz")['results'].item()
mean_df = get_mean_score(results)
mean_df.to_csv("./mean_df_converted.csv", index=False)
print(mean_df.shape)
mean_df.head(10)

(209, 8)


f1  accuracy  average_precision   roc_auc  precision    recall  \
0  0.080000  0.988705           0.408457  0.873482   0.400000  0.044444   
1  0.762071  0.689145           0.785920  0.742719   0.682989  0.864217   
2  0.000000  0.998618           0.168920  0.628251   0.000000  0.000000   
3  0.712121  0.591429           0.643214  0.411111   0.604286  0.900000   
4  0.000000  0.818182           0.222150  0.388889   0.000000  0.000000   
5  0.273907  0.784866           0.512232  0.736592   0.660476  0.175484   
6  0.064461  0.865974           0.391856  0.733942   0.866667  0.034040   
7  0.012698  0.930264           0.195145  0.671045   0.300000  0.006504   
8  0.502233  0.685106           0.675068  0.728110   0.657874  0.407571   
9  0.000000  0.992969           0.056756  0.633252   0.000000  0.000000   

   pos_num  neg_num  
0       45     3762  
1      383      283  
2       13     9395  
3       18       14  
4       10       45  
5      154      506  
6      383     2400  
7      309     4122  
8      297      462  
9       58     8333

In [14]:
print(("Across 209 assays, the average f1={:.2f}%, accuracy={:.2f}%, ap={:.2f}%, auc={:.2f}%, " + \
       "precision={:.2f}%, recall={:.2f}%.").format(
    np.mean(mean_df["f1"]) * 100,
    np.mean(mean_df["accuracy"]) * 100,
    np.mean(mean_df["average_precision"]) * 100,
    np.mean(mean_df["roc_auc"]) * 100,
    np.mean(mean_df["precision"]) * 100,
    np.mean(mean_df["recall"]) * 100))

Across 209 assays, the average f1=13.18%, accuracy=90.72%, ap=31.74%, auc=71.22%, precision=30.44%, recall=11.43%.


![](./plots/random_forest_plot_2.png)

### 1.2.1. Comment

We can see the same pattern as the random forest results from 141 assays. In conclusion, `AUC` is not a good metrics, and it is good that we have this fingerprint baseline.

## 2. Random Forest with Inception Feature

We can use the Inception V3 extracted features from U20S images to build end-to-end single task models. It can tell us more about our future image-to-assay model.

In this section, we will use the data with 209 assays. Due to the size limit of Gluster, we only have 244 plates. The paper reports that they have 375 plates, and their shared download script implies there are 406 plates. Therefore, we will not have features for all used compounds in the 209 assays.

In [5]:
# Load the output matrix
output_matrix_npz = np.load("./resource/output_matrix_convert.npz")
compound_inchi = output_matrix_npz["compound_inchi"]
compound_broad_id = output_matrix_npz["compound_broad_id"]
output_matrix = output_matrix_npz["output_matrix"]
output_matrix.shape

(26638, 209)

In [9]:
# It is slow to work with DataFrame, we can make a dictionary with BID -> PID+WID
mean_well_df = pd.read_csv("./resource/merged_mean_table.csv")

In [10]:
all_pids = mean_well_df["Metadata_Plate"].tolist()
all_wids = mean_well_df["Metadata_Well"].tolist()
all_bids = mean_well_df["Metadata_broad_sample"].tolist()

mean_well_dict = {}
intersect_bids = set(compound_broad_id)
for i in range(len(all_bids)):
    cur_bid = all_bids[i]
    if cur_bid in intersect_bids:
        if cur_bid not in mean_well_dict:
            mean_well_dict[cur_bid] = {"pid": [all_pids[i]],
                                       "wid": [all_wids[i]]}
        else:
            mean_well_dict[cur_bid]["pid"].append(all_pids[i])
            mean_well_dict[cur_bid]["wid"].append(all_wids[i])
     
print("From our 244 plates, we found {} compounds over {} total intersected compounds.".format(
    len(mean_well_dict), len(compound_broad_id)))

From our 244 plates, we found 17675 compounds over 26638 total intersected compounds.


In [23]:
# Reorder pids and wids to match output matrix row order
ordered_pids, ordered_wids = [], []
compound_replicate = {}

for b in compound_broad_id:
    if b in mean_well_dict:
        ordered_pids.extend(mean_well_dict[b]["pid"])
        ordered_wids.extend(mean_well_dict[b]["wid"])
        compound_replicate[b] = len(mean_well_dict[b]["pid"])

In [30]:
Counter(compound_replicate.values())

Counter({4: 14588, 3: 2095, 2: 534, 1: 457, 8: 1})

In our dataset, most compounds have 4 replicates (tested on 4 wells). For each well, we have 6-9 field of view shots, so we have about 24 samples for one compound.

In [47]:
# Extract features for those plates & wells
pwid = ["{}_{}".format(ordered_pids[i], ordered_wids[i]) for i in range(len(ordered_wids))]
pwid_set = set(pwid)

combined_feature = np.load("/Users/JayWong/Downloads/combined_feature.npz")
combined_feature_pwids = combined_feature["names"]

# Get matching indices
matched_indices = []
matched_pwids = []
for i in range(len(combined_feature_pwids)):
    if combined_feature_pwids[i] in pwid_set:
        matched_indices.append(i)
        matched_pwids.append(combined_feature_pwids[i])

In [61]:
all_feature = combined_feature["features"]
matched_feature = all_feature[matched_indices,:]
matched_names = combined_feature_pwids[matched_indices]
np.savez("./resource/matched_184_raw_features.npz", features=matched_feature, names=matched_names)

In [56]:
matched_features = np.load("./resource/matched_184_raw_features.npz")["features"]
matched_names = np.load("./resource/matched_184_raw_features.npz")["names"]
output_matrix = np.load("./resource/output_matrix_inception.npz")["output_matrix"]

In [46]:
# Rearrange output matrix to match our feature matrix
print(matched_features.shape)

pwd_to_b_dict = {}
for k, v in mean_well_dict.items():
    for i in range(len(v["pid"])):
        pwd = "{}_{}".format(v["pid"][i], v["wid"][i])
        if pwd in pwd_to_b_dict:
            print(pwd)
        pwd_to_b_dict[pwd] = k
        
output_compound_b_to_index = dict(zip(compound_broad_id, range(len(compound_broad_id))))

# Create the indices to select entries from output matrix
select_bids = []
select_indices = []
for pw in matched_names:
    bid = pwd_to_b_dict[pw]
    select_bids.append(bid)
    select_indices.append(output_compound_b_to_index[bid])

print("{} compounds found in our extracted features (we only extracted 184 plates).".format(len(set(select_bids))))

# Rearrange output matrix

output_matrix_npz = np.load("./resource/output_matrix_convert.npz")["output_matrix"]
rearranged_output_matrix = output_matrix[select_indices,:]
rearranged_output_matrix.shape
np.savez("./resource/output_matrix_inception.npz", output_matrix=rearranged_output_matrix)

(546690, 4096)
15019 compounds found in our extracted features (we only extracted 184 plates).


In [59]:
def train_lr_on_assay(aid, kfold=5, n_jobs=4):
    """
    (1). Use cross validation to find the best regularization parameter
    (2). Report the best parameter, and scores on the test set

    The model is only trained on compounds which give non-NA results.
    """
    assay_array = rearranged_output_matrix[:, aid]

    # Filter out NA from the assay_array
    y_index = [False if a == -1 else True for a in assay_array ]
    y = assay_array[y_index]
    x = matched_features[y_index, :]
    
    print(x.shape, y.shape)
    return
    
    if Counter(y)[1] < kfold:
        print("Warning: the number of total postives is less than kfold")
        
    
    # Check if the training samples are too small
    if Counter(y)[1] < 10 or Counter(y)[0] < 10:
        print("Not enough training samples for assay {}".formt(aid))
        return
    
    # Build the cross validation scheme
    # Since some assays have extremely small number of positives,
    # I will use StratifiedKFold to preserve the proportion of positive in test set
    
    my_kfold_gen = StratifiedKFold(n_splits=kfold, shuffle=True)
    scoring = ["f1", "accuracy", "precision", "recall", "average_precision", "roc_auc"]
    gs = GridSearchCV(LogisticRegression(penalty='l1', verbose=0,
                                         solver='liblinear'),
                      {'C': [0.01, 0.1, 1, 10]},
                      cv=my_kfold_gen,
                      scoring=scoring,
                      refit=False,
                      n_jobs=n_jobs, verbose=2)
    
    gs.fit(x, y)
    result = gs.cv_results_
    result["count"] = Counter(y)
    return result

It takes 12 hours to train 3 assays since we have too many features and samples ($546690 \times 4096$). We can use L1 logistic regression to select features first.

In [24]:
assay_array = output_matrix[:, 5].copy()

# Filter out NA from the assay_array
y_index = [False if a == -1 else True for a in assay_array ]
y = assay_array[y_index].copy()
x = matched_features[y_index, :].copy()

In [27]:
lr = LogisticRegression(penalty='l1', verbose=1, solver='saga', max_iter=1000)

In [28]:
lr.fit(x,y)

max_iter reached after 1018 seconds


/usr/local/lib/python3.6/site-packages/sklearn/linear_model/sag.py:326: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  num_seen_init,
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed: 17.0min finished


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=1000, multi_class='ovr', n_jobs=1,
          penalty='l1', random_state=None, solver='saga', tol=0.0001,
          verbose=1, warm_start=False)

In [58]:
abs_coef = np.abs(lr.coef_[0])
largest_coef_index = np.argpartition(abs_coef, -512)[-512:]
#abs_coef[largest_index]

In [62]:
assay_array = output_matrix[:, 0].copy()

# Filter out NA from the assay_array
y_index = [False if a == -1 else True for a in assay_array ]
y = assay_array[y_index].copy()

# Only choose selected 512 features
x = matched_features[y_index, :].copy()
x = x[:, largest_coef_index].copy()

In [65]:
matched_features = None
output_matrix = None

In [66]:
lr_2 = LogisticRegression(penalty='l1', verbose=1, solver='saga', max_iter=1000)
lr_2.fit(x,y)

KeyboardInterrupt: 

In [68]:
np.savez("feature_index.npz", index=largest_coef_index)